crates/data-store/src/lib.rs

```diff
@@ -14,13 +14,35 @@
  * limitations under the License.
  */
 
-/// This trait should be used to persist prev_data between successive calls of an interpreter o.
-pub trait DataStore<E> {
-    fn initialize(&mut self) -> Result<(), E>;
+use std::time::Duration;
 
-    fn store_data(&mut self, data: &[u8], key: &str) -> Result<(), E>;
+/// This trait is used for
+///   - persisting prev_data between successive calls of an interpreter
+///   - logging previous, current, and new data in case of spikes
+pub trait DataStore {
+    type Error;
 
-    fn read_data(&mut self, key: &str) -> Result<Vec<u8>, E>;
+    fn initialize(&mut self) -> Result<(), Self::Error>;
 
-    fn cleanup_data(&mut self, key: &str) -> Result<(), E>;
+    fn store_data(&mut self, data: &[u8], key: &str) -> Result<(), Self::Error>;
+
+    fn read_data(&mut self, key: &str) -> Result<Vec<u8>, Self::Error>;
+
+    /// Cleanup data that become obsolete.
+    fn cleanup_data(&mut self, key: &str) -> Result<(), Self::Error>;
+
+    /// Returns true if an anomaly happened and it's necessary to save execution data
+    /// for debugging purposes.
+    ///  execution_time - is a time taken by the interpreter to execute provided script
+    ///  memory_delta - is a count of bytes on which an interpreter heap has been extended
+    ///                 during execution of a particle
+    fn detect_anomaly(&self, execution_time: Duration, memory_delta: usize) -> bool;
```

Comment:
It might be bit disputable to put anomaly detection logic into a DataStore.

Reply:
I had the same thoughts and we dicsussed it with @folex and it looks like node code would become too complicated after introducing a new boxed trait to AVM. So we decided to do that in this trait.